### PR TITLE
feat(connector): [Adyen] Add support for PIX Payment Method

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
`{
    "amount": 6500,
    "currency": "BRL",
    "confirm": true,
    "capture_method": "manual",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "StripeCustomer",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "profile_id": "pro_sx78reJN0zgRzB5UGicx",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://duck.com",
    "billing": {
        "address": {
            "first_name": "John",
            "last_name": "Doe",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "BR"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "BR",
            "first_name": "John",
            "last_name": "Doe"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
        "payment_method": "bank_transfer",
    "payment_method_type": "pix",
    "payment_method_data": {
        "bank_transfer": {
            "pix": {}
        }
    }
}`
Screenshot 2024-01-23 at 6 24 59 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1894,6 +1898,34 @@ pub enum NextActionData {
...
+        bank_transfer_steps_and_charges_details: BankTransferNextStepsData,
+    },
+    QrCodeInformation(QrCodeInformation),
+}
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(untagged)]
+pub enum QrCodeInformation {
+    QrCodeUrl {
```
Comment (ArjunKarthik) [inline crates/api_models/src/payments.rs:1914]: Please add comments stating that the enum order shouldn't be changed as this is being used during serialization and deserialization

File: crates/router/src/connector/adyen/transformers.rs [hunk 2]
```diff
@@ -3325,14 +3297,22 @@ pub fn get_qr_metadata(
     let image_data_url = Url::parse(image_data.data.as_str())
         .ok()
         .ok_or(errors::ConnectorError::ResponseHandlingFailed)?;
+    let qr_code_url = response
+        .action
+        .qr_code_url
+        .clone()
+        .ok_or(errors::ConnectorError::ResponseHandlingFailed)?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:3304]: Why we are throwing error incase of None? This is an optional field from connector

File: crates/router/src/connector/adyen/transformers.rs [hunk 3]
```diff
@@ -3325,14 +3297,22 @@ pub fn get_qr_metadata(
...
 
-    let qr_code_instructions = payments::QrCodeNextStepsInstruction {
-        image_data_url,
-        display_to_timestamp: None,
-    };
+    let qr_code_instructions = payments::NextActionFromConnectorMetaData::QrCodeInformation(
+        payments::QrCodeInformation::QrCodeUrl {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:3307]: Based on the data available, please construct QrCodeInformation::QrCodeUrl or QrCodeInformation::QrDataUrl or QrCodeInformation::QrCodeImageUrl.

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -3325,14 +3297,22 @@ pub fn get_qr_metadata(
...
+            image_data_url,
+            qr_code_url,
+            display_to_timestamp: None,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:3310]: Don't we get expiration time limit for the QR code from Aden?

File: crates/router/src/core/payments/transformers.rs [hunk 5]
```diff
@@ -492,40 +492,59 @@ where
...
+                        .map(|next_action| match next_action {
+                            api_models::payments::NextActionFromConnectorMetaData::DisplayBankTransferInformation {
+                                bank_transfer_steps_and_charges_details } => {
+                                    api_models::payments::NextActionData::DisplayBankTransferInformation {
+                                        bank_transfer_steps_and_charges_details,
+                                    }
+                                },
+                                api_models::payments::NextActionFromConnectorMetaData::QrCodeInformation (qr_info) => {
+
+                                    match qr_info {
```
Comment (ArjunKarthik) [inline crates/router/src/core/payments/transformers.rs:518]: Move this to from implementation

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

File: crates/api_models/src/payments.rs [hunk 6]
```diff
@@ -1892,8 +1892,12 @@ pub enum NextActionData {
     /// Contains url for Qr code image, this qr code has to be shown in sdk
     QrCodeInformation {
         #[schema(value_type = String)]
-        image_data_url: Url,
+        /// Hyperswitch generated image data source url
+        image_data_url: Option<Url>,
         display_to_timestamp: Option<i64>,
+        #[schema(value_type = String)]
+        /// The url for Qr code given by the connector
+        qr_code_url: Option<Url>,
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/api_models/src/payments.rs:1900]: @sai-harsha-vardhan Both are 2 different URL. One field holds the redirect link where another one holds the data url

**Review by sai-harsha-vardhan — DISMISSED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**